src/List.mo

```diff
@@ -129,67 +158,84 @@ module {
   ///
   /// Space: `O(size)`
   public func fromPure<T>(pure : PureList.List<T>) : List<T> {
-    let list = empty<T>();
-    PureList.forEach<T>(pure, func(x) = add(list, x));
-    list
+    var p = pure;
+    var list = empty<T>();
+    loop {
+      switch (p) {
+        case (?(x, xs)) {
+          add(list, x);
+          p := xs
+        };
+        case null return list
+      }
+    }
```

Comment:
I would just keep the old code using `Pure.forEach` because that function is efficient. We are only adding one function call for each element (the wrapper around add()). `fromPure` is probably infrequent or a one-time operation.